src/sentry/replays/consumers/buffered/consumer.py

```diff
@@ -0,0 +1,128 @@
+"""Session Replay recording consumer implementation.
+
+To understand how the buffering works visit the `lib.py` module and inspect the source of the
+buffering runtime.
+
+This module has two parts. A processing component and a buffer flushing component. The processing
+component is straight-forward. It accepts a message and performs some work on it. After it
+completes it instructs the runtime to append the message to the buffer. This is abstracted by the
+buffering runtime library so we just return the transformed data in this module.
+
+The second part is the flushing of the buffer. The buffering runtime library has no idea when to
+flush this buffer so it constantly asks us if it can flush. We control flushing behavior through a
+stateful "BufferManager" class.  If we can_flush then we do_flush. After the flush completes the
+RunTime will commit the offsets.
+"""
+
+import contextlib
+import time
+from concurrent.futures import FIRST_EXCEPTION, ThreadPoolExecutor, wait
+from typing import TypedDict
+
+import sentry_sdk
+
+from sentry.replays.consumers.buffered.lib import Model, buffering_runtime
+from sentry.replays.usecases.ingest import (
+    DropSilently,
+    ProcessedRecordingMessage,
+    commit_recording_message,
+    parse_recording_message,
+    process_recording_message,
+    sentry_tracing,
+    track_recording_metadata,
+)
+
+
+class Flags(TypedDict):
+    max_buffer_length: int
+    max_buffer_wait: int
+    max_workers: int
+
+
+class BufferManager:
+    """Buffer manager.
+
+    The buffer manager is a class instance has a lifetime as long as the RunTime's. We pass its
+    methods as callbacks to the Model. The state contained within the method's instance is implicit
+    and unknown to the RunTime.
+    """
+
+    def __init__(self, flags: Flags) -> None:
+        self.__max_buffer_length = flags["max_buffer_length"]
+        self.__max_buffer_wait = flags["max_buffer_wait"]
+        self.__max_workers = flags["max_workers"]
+
+        self.__last_flushed_at = time.time()
+
+    def can_flush(self, model: Model[ProcessedRecordingMessage]) -> bool:
+        # TODO: time.time is stateful and hard to test. We should enable the RunTime to perform
+        #       managed effects so we can properly test this behavior.
+        return (
+            len(model.buffer) >= self.__max_buffer_length
+            or (time.time() - self.__max_buffer_wait) >= self.__last_flushed_at
+        )
```

Comment:
Arroyo primitives manage these kind of concerns for you (when to flush a batch for example). Are you sure about the idea of pushing them into the product code instead ?